### PR TITLE
feat: add Menu component

### PR DESCRIPTION
**Pull request by austinhung81**
# Menu

An accessible dropdown menu for the common dropdown menu button design pattern.
Menu uses roving tabIndex for focus management.

<carbon-ad></carbon-ad>

## Import

- `Menu`: The wrapper component that provides context, state, and focus
  management.
- `MenuList`: The wrapper for the menu items. Must be a direct child of `Menu`.
- `MenuItem`: The trigger that handles menu selection. Must be a direct child of
  a `MenuList`.
- `MenuGroup`: A wrapper to group related menu items.
- `MenuDivider`: A visual separator for menu items and groups.

```js
import { Menu, MenuList, MenuItem, MenuGroup, MenuDivider } from '@trendmicro/react-styled-ui';
// or
import Menu from '@trendmicro/react-styled-ui/Menu';
```

- `MenuButton`: The trigger for the menu list. Must be a direct child of `Menu`.

```js
import MenuButton from '@trendmicro/react-styled-ui';
```

### Basic menu with an anchor

```jsx
function basicMenu() {
  const [anchorEl, setAnchorEl] = React.useState(null);
  const { isOpen, onClose, onToggle } = useDisclosure();

  const handleClick = (event) => {
    setAnchorEl(event.currentTarget);
    onToggle();
  };

  return (
    <>
      <Button onClick={handleClick}>Select an option</Button>
      <Menu
        anchorEl={anchorEl}
        isOpen={isOpen}
        onClose={onClose}
      >
        <MenuList>
          <MenuItem>List item</MenuItem>
          <MenuItem>List item</MenuItem>
          <MenuItem>List item</MenuItem>
        </MenuList>
      </Menu>
    </>
  );
}
```

### Basic menu with `MenuButton`

```jsx
<Stack shouldWrapChildren direction="row" spacing="2x">
  <Menu>
    <MenuButton>Select an option</MenuButton>
    <MenuList>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuList>
  </Menu>
  <Menu>
    <MenuButton variant="ghost">Select an option</MenuButton>
    <MenuList>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuList>
  </Menu>
</Stack>
```

### Group

```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList>
    <MenuGroup title="title1">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
    <MenuGroup title="title2">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
  </MenuList>
</Menu>
```

### Divider

```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList>
    <MenuGroup title="title1">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
    <MenuDivider />
    <MenuGroup title="title2">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
  </MenuList>
</Menu>
```

### Disabled
Add `disabled` property on `MenuItem`

```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList>
    <MenuGroup title="title1">
      <MenuItem>List item</MenuItem>
      <MenuItem disabled>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
    <MenuDivider />
    <MenuGroup title="title2">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem disabled>List item</MenuItem>
    </MenuGroup>
  </MenuList>
</Menu>
```

### Set the minimum width of the `MenuList` to the width of the `MenuButton`

```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList minWidth="100%">
    <MenuGroup title="title1">
      <MenuItem>List item</MenuItem>
      <MenuItem disabled>List item</MenuItem>
      <MenuItem>List item</MenuItem>
    </MenuGroup>
    <MenuDivider />
    <MenuGroup title="title2">
      <MenuItem>List item</MenuItem>
      <MenuItem>List item</MenuItem>
      <MenuItem disabled>List item</MenuItem>
    </MenuGroup>
  </MenuList>
</Menu>
```

### Menu with icon

```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList>
    <MenuItem><TMIcon name="delete" mr="2x"/>List item</MenuItem>
    <MenuItem><TMIcon name="edit" mr="2x"/>List item</MenuItem>
    <MenuItem><TMIcon name="share" mr="2x"/>List item</MenuItem>
  </MenuList>
</Menu>
```

### Menu with checkbox
Need to set `closeOnSelect` to `false` on `Menu`
```jsx
function menuWithCheckbox() {
  const [checkedItems, setCheckedItems] = React.useState(['host', 'ip']);
  const handleChange = (e) => {
    const value = e.target.value;
    const items = [...checkedItems];
    items.indexOf(value) === -1 ? items.push(value) : items.splice(items.indexOf(value), 1);
    setCheckedItems(items);
  };
  return (
    <Menu closeOnSelect={false}>
      <MenuButton>Select an option</MenuButton>
      <MenuList>
        <CheckboxGroup size="sm" defaultValue={checkedItems}>
          <MenuItem><Checkbox value="host" onChange={e => handleChange(e)}>Endpoint name</Checkbox></MenuItem>
          <MenuItem><Checkbox value="ip" onChange={e => handleChange(e)}>IP address</Checkbox></MenuItem>
          <MenuItem><Checkbox value="os" onChange={e => handleChange(e)}>Operating system</Checkbox></MenuItem>
        </CheckboxGroup>
      </MenuList>
    </Menu>
  );
}
```

### Menu with label inline

```jsx
function inlineLabelMenu() {
  const [menuItem, setMenuItem] = React.useState('Name');

  const handleClick = (event) => {
    if (event.key === 'Enter' || event.type === 'click') {setMenuItem(event.target.innerHTML)}
  };

  return (
    <Menu>
      <MenuButton><Text color="white:secondary" mr="1x">Search by:</Text>{menuItem}</MenuButton>
      <MenuList onClick={handleClick} minWidth="100%">
        <MenuItem onKeyDown={handleClick}>Name</MenuItem>
        <MenuItem onKeyDown={handleClick}>Address</MenuItem>
        <MenuItem onKeyDown={handleClick}>Postcode</MenuItem>
      </MenuList>
    </Menu>
  );
}
```

### Offset position
Add `skidding` or `distance` to customize menu offset position
```jsx
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList skidding={20} distance={8}>
    <MenuItem>List item</MenuItem>
    <MenuItem>List item</MenuItem>
    <MenuItem>List item</MenuItem>
  </MenuList>
</Menu>
```

### Keyboard Interaction

| Key                | Action                                                                                                                                  |
| ------------------ | --------------------------------------------------------------------------------------------------------------------------------------- |
| `Enter` or `Space` | When `MenuButton` receives focus, opens the menu and places focus on the first menu item                                                |
| `ArrowDown`        | When `MenuButton` receives focus, opens the menu and moves focus to the first menu item                                                 |
| `ArrowUp`          | When `MenuButton` receives focus, opens the menu and moves focus to the last menu item                                                  |
| `Escape`           | When the menu is open, closes the menu and sets focus to the `MenuButton`                                                               |

### ARIA roles

**For `MenuButton`:**

- It has `role` set to `button`.
- It has aria-haspopup set to either `menu`.
- When the menu is displayed, `MenuButton` has `aria-expanded` set to `true`.
- `MenuButton` has `aria-controls` set to the `id` of the `MenuList`.

**For `MenuList`:**

- It contains the `MenuItem` has role `menu`.

### Menu Props

| Name          | Type               | Default | Description                                                  |
| ------------- | ------------------ | ------- | ------------------------------------------------------------ |
| children      | React.ReactNode    |         | The children of the menu must be `MenuButton` or `MenuList`  |
| isOpen        | boolean            |         | If `true`, the menu will be opened                           |
| autoSelect    | boolean            | `false` | The menu will select the first enabled item when it opens, the trigger item must be `MenuButton`    |
| closeOnBlur   | boolean            | `true`  | If `true`, the menu will close on outside click or blur      |
| closeOnSelect | boolean            | `true`  | If `true`, the menu will close on menu item select           |
| placement     | PopperJS.placement | `bottom-start` | The placement of the `MenuList`. One of: `'top'`, `'bottom'`, `'right'`, `'left'`, `'top-start'`, `'top-end'`, `'bottom-start'`, `'bottom-end'`, `'right-start'`, `'right-end'`, `'left-start'`, `'left-end'`  |

### MenuButton Props

| Name      | Type                       |
| --------- | -------------------------- |
| children  | React.ReactNode            |
| onClick   | React.MouseEventHandler    |
| onKeyDown | React.KeyboardEventHandler |

### MenuList Props

| Name      | Type                    | Default        | Description                                                       |
| --------- | ----------------------- | -------------- | ----------------------------------------------------------------- |
| children  | React.ReactNode         |                | The content of the `MenuList`, should be the `MenuItem` component |
| onClick   | React.MouseEventHandler |                | Function to be called when you click on the menu item             |
| onBlur    | React.FocusEventHandler |                | Function to be called when you blur out of the menu list          |
| skidding  | number                  | `0`            | Displaces the menu (in pixels) along the reference element. Used by [`popper.js`](https://popper.js.org/docs/v2/modifiers/offset/#skidding-1)          |
| distance  | number                  | `0`            | Displaces the menu (in pixels) away from, or toward, the reference. Used by [`popper.js`](https://popper.js.org/docs/v2/modifiers/offset/#distance-1)  |

### MenuItem Props

| Name       | Type                                          | Description                                               |
| ---------- | --------------------------------------------- | --------------------------------------------------------- |
| disabled   | boolean                                       | If `true`, the menu item will be disabled                 |
| onClick    | React.MouseEventHandler                       | Function that is called on click and enter/space keypress |
| onKeyDown  | React.KeyboardEventHander                     | Function that is called on keydown                        |

### MenuGroup Props

| Name     | Type            | Description                   |
| -------- | --------------- | ----------------------------- |
| children | React.ReactNode | The content of the menu group, should be the `MenuItem` component |
| title    | string          | The title of the menu group   |

### CODE REVIEW

File: packages/react-styled-ui/src/Divider/index.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      aria-orientation={orientation}
+      border="0"
+      {...borderProps}
+      borderColor={borderColor}
+      {...props}
+    />
+  );
+});
+
+export default Divider;
```
Comment (roth1002) [inline packages/react-styled-ui/src/Divider/index.js:29]: Divider的Spec在哪裡啊？

File: packages/react-styled-ui/src/Menu/styles.js [hunk 2]
```diff
@@ -29,48 +50,53 @@ export const useMenuListStyle = () => {
...
-  display: 'flex',
-  textDecoration: 'none',
-  alignItems: 'center',
-  textAlign: 'left',
-  outline: 'none'
+const baseProps = ({ isMenuGrouped }) => {
+  return {
+    flex: ' 0 0 auto',
+    userSelect: 'none',
+    px: isMenuGrouped ? '6x' : '3x',
```
Comment (roth1002) [inline packages/react-styled-ui/src/Menu/styles.js]: isMenuGrouped 相當於是一種 variantProps的概念

如果要在這邊處理就失去 baseProps這個命名了

File: packages/react-styled-ui/src/MenuButton/index.js [hunk 3]
```diff
@@ -62,19 +67,18 @@ const MenuButton = forwardRef(
...
-          >
-            <Icon width="4x" name="_core.angle-down" />
-          </PseudoBox>
-        )}
-      </Comp>
+        <PseudoBox
+          aria-disabled={disabled}
+          {...iconWrapperProps}
+        >
+          <Icon width="4x" name="_core.angle-down" />
```
Comment (roth1002) [inline packages/react-styled-ui/src/MenuButton/index.js]: 這個 Icon 要客製化的寫法可以看一下 Alert/AlertToast那邊

File: packages/react-styled-ui/src/Divider/index.js (hunk 1 above)
Comment (austinhung81) [reply packages/react-styled-ui/src/Divider/index.js:29]: Divider 的spec 尚未出來，所以我先用了Menu 的spec, 我會在詢問Pin Pin，我沒將它寫成mdx 就是因為不確定他的spec

File: packages/react-styled-ui/src/Menu/styles.js (hunk 2 above)
Comment (austinhung81) [reply packages/react-styled-ui/src/Menu/styles.js]: 這個還在討論～這就攸關，你認為有需要幫user 縮排嗎？

File: packages/react-styled-ui/src/MenuButton/index.js (hunk 3 above)
Comment (austinhung81) [reply packages/react-styled-ui/src/MenuButton/index.js]: okok我來看看

Comment (austinhung81) [reply packages/react-styled-ui/src/MenuButton/index.js]: 我留一些後路讓他可以依照menu 的 placement 來變更icon 方向
fixed in 7f49e11

File: packages/react-styled-ui/src/Menu/MenuGroup.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  return (
+    <>
+      <Box
+        ref={ref}
+        role="group"
+        {...styleProps}
+      >
+        {title}
+      </Box>
+      <MenuGroupContextProvider value={{}}>
```
Comment (tinaClin) [inline packages/react-styled-ui/src/Menu/MenuGroup.js]: 使用`MenuGroupContextProvider`傳空物件，是為了判斷`<MenuItem>`是不是在`<MenuGroup>` 裡(要把Group裡的子項目做縮排)，我不確認這是不是好的作法。 @cheton @roth1002 有什麼建議嗎？

Comment (roth1002) [reply packages/react-styled-ui/src/Menu/MenuGroup.js]: 這目前我沒有意見XD

File: packages/react-styled-ui/src/Divider/index.js (hunk 1 above)
Comment (austinhung81) [reply packages/react-styled-ui/src/Divider/index.js:29]: Pin 說先follow menu spec 做

File: packages/react-styled-ui/src/Menu/MenuItem.js [hunk 5]
```diff
@@ -24,12 +24,13 @@ const MenuItem = forwardRef(
...
-
-    const styleProps = useMenuItemStyle();
+    const menuGroupContext = useMenuGroup();
+    const isMenuGrouped = !!menuGroupContext;
+    const styleProps = useMenuItemStyle({ isMenuGrouped });
 
     return (
       <PseudoBox
-        as="li"
+        as="div"
```
Comment (cheton) [inline packages/react-styled-ui/src/Menu/MenuItem.js]: `as` 在這可以不用寫

Comment (austinhung81) [reply packages/react-styled-ui/src/Menu/MenuItem.js]: 阿因為原本用li哈哈好，我改個🙏

File: packages/react-styled-ui/src/Menu/MenuList.js [hunk 6]
```diff
@@ -78,12 +78,12 @@ const MenuList = ({ ...props }) => {
 
   return (
     <Popper
-      as="ul"
+      as="div"
```
Comment (cheton) [inline packages/react-styled-ui/src/Menu/MenuList.js]: 這裡的 `as` 應該也可以拿掉了

File: packages/react-styled-ui/src/Menu/styles.js (hunk 2 above)
Comment (cheton) [reply packages/react-styled-ui/src/Menu/styles.js]: 他的意思是說，如果需要處理縮排就是多一個 `menuGroupProps`，不要在 `baseProps` 做這件事情

```js
  return {
    ...baseProps(props),
    ...menuGroupProps(props),
    ...menuItemProps(props),
  };	  
};
```

Comment (austinhung81) [reply packages/react-styled-ui/src/Menu/styles.js]: 直接拿掉摟 fixed in 2980c58

File: packages/react-styled-ui/src/Menu/MenuGroup.js (hunk 4 above)
Comment (austinhung81) [reply packages/react-styled-ui/src/Menu/MenuGroup.js]: 拿掉了 fixed in 2980c58

**Review by cheton — APPROVED** [review]

**Review by roth1002 — APPROVED** [review]

**Pull request merged**